### ISSUE TITLE
ACIQ clipping not in sync with paper

### ISSUE THREAD

**Issue by amitsrivastava78** [submission2019/cnn-quantization#5]
Hi @ynahshan , @submission2019,
            I checked your paper and code, so as per my understanding the after forward pass from each operator (example conv), there will be ACIQ based on the laplase (the input i had used for 4-bit quantizing the resnet50). So i assumed that the clipping will only clip the regions from left and right of the tensors, but when i plot the histogram i can see entirly different histogram for output, I am attaching one the histogram as reference to you, the Red line shows the input tensor and black one shows the output(clipped tensor).
Can you please explain the result which is not in sync with your paper or at least my understanding of your paper.

Regards
Amit 

![Clipping](https://user-images.githubusercontent.com/30853054/59507853-e7339980-8ec9-11e9-82e9-1a09e8b3caaa.png)

**Comment by amitsrivastava78** [comment]
Also there is another histogram from the same flow which show slight traces but still not same as per the paper.
Attaching the same below as mentioned above red is the input tensor after the opeator forward pass and black line represents the ACIQ tensor.

![Clipping1](https://user-images.githubusercontent.com/30853054/59508062-5610f280-8eca-11e9-966d-8220d9a9645c.png)


Can you please let me know why i am i seeing this kind of behaviour

Regards
Amit

**Comment by submission2019** [comment]
Hi @amitsrivastava78.
ACIQ clipping applying as part of quantization procedure. In order to observe it's effect on tensor level you should dump data before and after quantization. In case of convolution it could be done in inference_quantization_manager.py line 208 like this.
`
torch.save(out, os.path.join('dump', activation_id + '.pt'))

out = QMI().quantize_instant(out, tag_act, half_range=hasattr(self, 'before_relu'), verbose=QMI().verbose)

torch.save(out, os.path.join('dump', activation_id + '_q.pt'))
`

Also note that quantization and ACIQ applied on every channel separately. Here is some histograms that I did on arbitrary layers and channels.
IMHO it is hard to see the effect of ACIQ on histogram and it is better look into statistics of all layers of min/max with and w/o ACIQ.

![image](https://user-images.githubusercontent.com/43610359/59561093-7d7bd280-9024-11e9-89fc-4e179f109422.png)
![image](https://user-images.githubusercontent.com/43610359/59561097-89679480-9024-11e9-8ef1-3b19e5589c33.png)
![image](https://user-images.githubusercontent.com/43610359/59561135-0f83db00-9025-11e9-8e07-0d7095112815.png)
![image](https://user-images.githubusercontent.com/43610359/59561137-16125280-9025-11e9-9018-cfa953d2a006.png)

**Comment by amitsrivastava78** [comment]
@submission2019 , @ynahshan , thanks for the response, i had put the code exactly where you asked me to put but still i did not observe proper ACIQ, rather the histogram came exactly what i had posted in my first two queries, please find the code snippet and histogram as below : -

![image](https://user-images.githubusercontent.com/30853054/59576071-d8bccc00-90db-11e9-87b5-a0ee2ff68b15.png)





![Clipping2](https://user-images.githubusercontent.com/30853054/59576087-fbe77b80-90db-11e9-9f94-522969b47e54.png)


Kindly help me to understand why is this happening ? Am i missing something here ?

Regards
Amit

**Comment by submission2019** [comment]
Hi. 
There is few problems with you code.
1. torch.histc() returns only counts of elements per bin with out any adges information so X axis of your plot always 50 with regards to the range of tensor.
2. Even with plt.hist() it is hard to see the range due to large amount of elements.

If you really want to measure range due to clipping the best way is just to print it like this:
`print(out.max() - out.min())`

But make sure you handle relu in cases where quantization is following relu.
```
if hasattr(self, 'before_relu'):
    out = torch.nn.functional.relu(out)
```

**Comment by submission2019** [comment]
Hi @amitsrivastava78. Do you have any more questions related to this topic? If not I will close this issue.
